src/twisted/internet/_idna.py

```diff
@@ -0,0 +1,54 @@
+# -*- test-case-name: twisted.test.test_sslverify -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+Shared interface to IDNA encoding and decoding, using the C{idna} PyPI package
+if available, otherwise the stdlib implementation.
+"""
+
+def _idnaBytes(text):
+    """
+    Convert some text typed by a human into some ASCII bytes.
+
+    This is provided to allow us to use the U{partially-broken IDNA
+    implementation in the standard library <http://bugs.python.org/issue17305>}
+    if the more-correct U{idna <https://pypi.python.org/pypi/idna>} package is
+    not available; C{service_identity} is somewhat stricter about this.
+
+    @param text: A domain name, hopefully.
+    @type text: L{unicode}
+
+    @return: The domain name's IDNA representation, encoded as bytes.
+    @rtype: L{bytes}
+    """
+    try:
+        import idna
```

Comment:
Does it make sense to instead do:

```python
try:
    from idna import encode as idna_encode, decode as idna_decode
except ImportError:
    def idna_encode(str):
        return str.encode("idna")
    def idna_decode(octets):
        return octects.decode("idna")
```

At the module level, and then just do:

```python
def _idnaBytes(text):
    return idna_encode(text)

def _idnaText(octects):
    return idna_decode(octects)
```

That way the import conditional stuff happens once at module scope instead of each time these functions are called? Particularly since IIRC missing imports are not cached like successful imports are.